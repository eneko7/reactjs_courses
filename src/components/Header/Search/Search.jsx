import React, { Component } from 'react';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import styles from './Search.scss';

class Search extends Component {
  searchInput = React.createRef();

  saveSearchRequest = () => {
    const { updateSearchRequest, history, location } = this.props;
    const parsed = queryString.parse(location.search);
    updateSearchRequest(this.searchInput.current.value);
    history.push(`/search?query=${this.searchInput.current.value}&searchBy=${parsed.searchBy}&sortBy=${parsed.sortBy}`);
  };

  searchHandler = (evenet) => {
    evenet.preventDefault();
    const {
      fetchSearchMovies,
      sortBy,
      searchBy,
      history,
    } = this.props;
    fetchSearchMovies(this.searchInput.current.value, searchBy, sortBy);
    if (this.searchInput.current.value !== '') {
      history.push(`/search?query=${this.searchInput.current.value}&searchBy=${searchBy}&sortBy=${sortBy}`);
    } else {
      history.push('');
    }
  };

  render() {
    const { searchRequest, searchBy, location } = this.props;
    const parsed = queryString.parse(location.search);
    let placeholder = 'Enter movie title';
    if ((parsed.searchBy && parsed.searchBy === 'genres') || searchBy === 'genres') {
      placeholder = 'Enter genres name, for example: action, family, adventure';
    }
    return (
      <form onSubmit={this.searchHandler}>
        <input
          type="text"
          className={styles.search_input}
          ref={this.searchInput}
          name="search"
          placeholder={placeholder}
          value={parsed.query ? parsed.query : searchRequest}
          onChange={this.saveSearchRequest}
        />
        <span className={styles.search_icon} onClick={this.searchHandler} role="presentation">&#8617;</span>
      </form>
    );
  }
}

Search.defaultProps = {
  sortBy: 'vote_average',
  searchBy: 'title',
  searchRequest: '',
};

Search.propTypes = {
  sortBy: PropTypes.string,
  searchBy: PropTypes.string,
  searchRequest: PropTypes.string,
  fetchSearchMovies: PropTypes.func.isRequired,
  updateSearchRequest: PropTypes.func.isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired,
  location: PropTypes.shape({
    search: PropTypes.string,
  }).isRequired,
};

export default Search;
