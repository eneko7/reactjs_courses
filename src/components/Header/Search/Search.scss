@import "../../../assets/constants/constants.scss";

.search_input {
  height: 50px;
  width: 100%;
  background-color: #000;
  position: relative;
  border: none;
  border-bottom: 2px solid $roseColor;
  z-index: 3;
  color: white;
  font-size: 20px;
  padding-left: 30px;
}

.search_input:focus {
  outline: none;
}

.search_input::placeholder {
  color: #FFF;
  opacity: 0.8;
}

.search_icon {
  color: #FFF;
  position: absolute;
  z-index: 3;
  right: 15%;
  font-size: 30px;
  width: 40px;
  line-height: 50px;
  cursor: pointer;
}

@media (max-width: $fulhdDesktop) {
  .search_icon {
    right: $leftPaddingLaptop;
  }
}

@media (max-width: $laptop) {
  .search_icon {
    right: $leftPaddingTablet;
  }
}

@media (max-width: $tablet) {
  .search_input {
    font-size: 16px;
  }

  .search_icon {
    right: 5%;
  }
}