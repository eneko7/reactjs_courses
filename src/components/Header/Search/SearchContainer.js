import { connect } from 'react-redux';
import Search from './Search';
import { fetchSearchMovies } from '../../../modules/movies/moviesActions';
import { getSearchBy, getSortBy, getSearchRequest } from '../../../modules/filter/filterSelectors';
import { updateSearchRequest } from '../../../modules/filter/filterActions';

const mapStateToProps = state => ({
  sortBy: getSortBy(state),
  searchBy: getSearchBy(state),
  searchRequest: getSearchRequest(state),
});

const mapDispatchToProps = {
  fetchSearchMovies,
  updateSearchRequest,
};
export default connect(mapStateToProps, mapDispatchToProps)(Search);
