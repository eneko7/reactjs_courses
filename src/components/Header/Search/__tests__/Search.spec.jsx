import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import Search from '../Search';

const shallow = new ShallowRenderer();
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const data = {
  sortBy: 'release_date',
  searchBy: 'title',
  searchRequest: 'test',
  fetchSearchMovies: () => ('Hello'),
  updateSearchRequest: () => ('Hello'),
};

const dataForGenres = {
  sortBy: 'release_date',
  searchBy: 'genres',
  searchRequest: 'test',
  fetchSearchMovies: () => ('Hello'),
  updateSearchRequest: () => ('Hello'),
};

const store = mockStore({
  sortBy: 'release_date',
  searchBy: 'title',
  searchRequest: 'test',
});

describe('Search render', () => {
  it('renders with title correctly', () => {
    const component = shallow.render(
      <Provider store={store}>
        <Search {...data} />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
  it('renders with genres correctly', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <Search {...dataForGenres} />
      </Provider>,
      {
        createNodeMock: (element) => {
          if (element.type === 'input') {
            return {
              value: 'Action',
            };
          }
          return null;
        },
      },
    );
    const event = { stopPropagation: jest.fn(), preventDefault: jest.fn() };
    component.root.findByProps({ className: 'search_input' }).props.onChange(event);
    expect(component).toMatchSnapshot();
  });
  it('change input value correctly', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <Search {...data} />
      </Provider>,
      {
        createNodeMock: (element) => {
          if (element.type === 'input') {
            return {
              value: 'word',
            };
          }
          return null;
        },
      },
    );
    const event = { stopPropagation: jest.fn(), preventDefault: jest.fn() };
    component.root.findByProps({ className: 'search_input' }).props.onChange(event);
    expect(component).toMatchSnapshot();
  });

  it('click on icon correctly', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <Search {...data} />
      </Provider>,
      {
        createNodeMock: (element) => {
          if (element.type === 'input') {
            return {
              value: 'word',
            };
          }
          return null;
        },
      },
    );
    const event = { stopPropagation: jest.fn(), preventDefault: jest.fn() };
    component.root.findByProps({ className: 'search_icon' }).props.onClick(event);
    expect(component).toMatchSnapshot();
  });
});
