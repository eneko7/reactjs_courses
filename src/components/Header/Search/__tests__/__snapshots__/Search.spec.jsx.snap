// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search render change input value correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    className="search_input"
    name="search"
    onChange={[Function]}
    placeholder="Enter movie title"
    type="text"
    value="test"
  />
  <span
    className="search_icon"
    onClick={[Function]}
    role="presentation"
  >
    ↩
  </span>
</form>
`;

exports[`Search render click on icon correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    className="search_input"
    name="search"
    onChange={[Function]}
    placeholder="Enter movie title"
    type="text"
    value="test"
  />
  <span
    className="search_icon"
    onClick={[Function]}
    role="presentation"
  >
    ↩
  </span>
</form>
`;

exports[`Search render renders with genres correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    className="search_input"
    name="search"
    onChange={[Function]}
    placeholder="Enter genres name, for example: action, family, adventure"
    type="text"
    value="test"
  />
  <span
    className="search_icon"
    onClick={[Function]}
    role="presentation"
  >
    ↩
  </span>
</form>
`;

exports[`Search render renders with title correctly 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeState": Object {
        "searchBy": "title",
        "searchRequest": "test",
        "sortBy": "release_date",
      },
    }
  }
>
  <Search
    fetchSearchMovies={[Function]}
    searchBy="title"
    searchRequest="test"
    sortBy="release_date"
    updateSearchRequest={[Function]}
  />
</Context.Provider>
`;
